compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1115,6 +1116,83 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         }
     }
 
+    /// Visit every module reachable from `start_module` and bring any proc and derive macro
+    /// encountered into `self.macro_map` to be used for suggestions.
+    fn lookup_macros_from_module(&mut self) {
```

Comment:
How is this different from `lookup_import_candidates_from_module`?

Only macros that would be suggested as imports (and their helper attributes) can be suggested here.
So I suggest removing this separate logic and retrieve helper attributes in the `lookup_import_candidates_from_module` pass if the import candidate happens to be a derive macro and helper attributes satisfy `filter_fn`.

Same for typo suggestions, in addition for checking any visited item for being typo-suggestable we also need to check its helper attributes for the same thing, as a part of the regular `early_lookup_typo_candidate` pass.

Reply:
I suggest landing the core change first - extending the set of regular import and typo suggestions with helper attributes, and leave all the beautifications for later.